Kernel/Arch/aarch64/PCI/Controller/BCM2712HostController.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2024-2025, Sönke Holz <sholz8530@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Arch/Delay.h>
+#include <Kernel/Boot/CommandLine.h>
+#include <Kernel/Bus/PCI/Access.h>
+#include <Kernel/Bus/PCI/DeviceTreeHelpers.h>
+#include <Kernel/Firmware/DeviceTree/DeviceTree.h>
+#include <Kernel/Firmware/DeviceTree/Driver.h>
+#include <Kernel/Firmware/DeviceTree/Management.h>
+#include <Kernel/Memory/TypedMapping.h>
+
+namespace Kernel::PCI {
+
+// This driver requires the host controller to be already initialized by the firmware.
+
+// This host controller is not ECAM-compliant.
+// The config space is accessible through a single 4K window that can be mapped to any bus/device/function.
+// The b/d/f can be configured by writing an ECAM offset to Registers::config_space_window_address.
+
+struct Registers {
+    u8 unknown1[0x400c];
+
+    // Bus base [31:0] @ [31:0]
+    u32 bus_window_base_low; // 0x400c
+
+    // Bus base [63:32] @ [31:0]
+    u32 bus_window_base_high; // 0x4010
+
+    u8 unknown2[0x4064 - (0x4010 + 4)];
+
+    u32 control; // 0x4064
+    static constexpr u32 CONTROL_PERST_N = 1 << 2;
+
+    u32 state; // 0x4068
+    static constexpr u32 STATE_LINK_UP = 0b11 << 4;
+
+    u8 unknown3[0x4070 - (0x4068 + 4)];
+
+    // The host controller doesn't seem to support 16-bit accesses, so the base and limit are grouped into one field.
+    // CPU limit [31:20] @ [31:20]
+    // CPU base [31:20] @ [15:4]
+    u32 cpu_window_base_low_and_cpu_window_limit_low; // 0x4070
```

Comment:
IMO we could use bitfields here which makes accessing and checking a bit easier (but writing iffier with RWC1)
but I think we had the discussion a few times

Maybe use bit-enums for the status (and maybe control) though

Reply:
Well that's a quick fix, just add virtual overloads to the helper xD

Reply:
Adding volatile overloads sadly isn't possible, as gcc throws this (nameless) warning if you ignore the result of `operator|=`:
```
/home/s/repos/serenity/Kernel/Arch/aarch64/PCI/Controller/BCM2712HostController.cpp:132:24: error: implicit dereference will not access object of type 'volatile Kernel::PCI::Registers::Control' in statement [-Werror]
  132 |     registers->control |= Registers::Control::PERST_N;
      |     ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```